File: backend/src/python/adapters/valuation.ts
/**
 * valuation adapter — hybrid runner.
 *
 * Ported from python-services/src/financex/calculators/valuation.py.
 * The DCF arithmetic itself already runs inside financial_engine
 * when valuation_inputs are supplied there; this adapter reshapes
 * whatever the engine produced (dcf block inside engine_snapshot),
 * layers peer multiple stats from sector_competition output, and
 * adds the three sector hint flags (TRY WACC warning, holding SOTP
 * required, banking FCF-DCF incompatible).
 *
 * Bull/Base/Bear narrative, SOTP composition, WACC policy choice,
 * scenario weighting, sensitivity matrix narration — all stay with
 * the LLM. This is the numeric skeleton only.
 *
 * When upstream financial_analysis is LLM markdown (mixed-flag mode),
 * we fall back to llm_fallback.resolveSector() to still get the
 * sector-aware flags right.
 */

import { resolveSector } from './llm_fallback.js';

export interface UpstreamRatioValue {
  value?: string | number | null;
  label?: string | null;
  warning?: string | null;
}

export interface UpstreamDcfResult {
  enterprise_value?: string | number | null;
  equity_value?: string | number | null;
  per_share_value?: string | number | null;
  wacc_used?: string | number | null;
  terminal_growth?: string | number | null;
  sensitivity?: Array<{ wacc?: unknown; terminal_g?: unknown; per_share?: unknown }>;
}

export interface UpstreamEngineSnapshot {
  dcf?: UpstreamDcfResult | null;
  ratios?: unknown;
  scores?: unknown;
}

export interface UpstreamFinancialAnalysis {
  ticker?: string;
  period_label?: string;
  sector?: string;
  engine_snapshot?: UpstreamEngineSnapshot | null;
  engine?: UpstreamEngineSnapshot | null;
}

export interface UpstreamSectorCompetition {
  benchmarks?: Array<{
    metric_code: string;
    company_value?: number | null;
    median?: number | null;
    q1?: number | null;
    q3?: number | null;
  }>;
}


export interface PeerMultipleStats {
  metric: string;
  count: number;
  median: number | null;
  q1: number | null;
  q3: number | null;
  company_value: number | null;
  company_vs_median: number | null;
}


export interface LegacyValuationOutput {
  agent_id: string;
  output_id: string;
  ticker: string;
  period_label: string;
  sector: string;
  dcf: UpstreamDcfResult | null;
  peer_ev_ebitda: PeerMultipleStats | null;
  peer_pe: PeerMultipleStats | null;
  try_wacc_warning: boolean;
  holding_sotp_required: boolean;
  banking_sector_warning: boolean;
  sector_source: 'structured' | 'markdown' | 'ticker';
  notes: string[];
  warnings: string[];
  review_status: string;
  source: 'python';
}


function unwrap<T>(raw: unknown): T | null {
  if (raw == null) return null;
  if (typeof raw === 'string') {
    try { return JSON.parse(raw) as T; } catch { return null; }
  }
  return typeof raw === 'object' ? (raw as T) : null;
}


export function extractFinancialAnalysis(raw: unknown): UpstreamFinancialAnalysis | null {
  return unwrap<UpstreamFinancialAnalysis>(raw);
}


export function extractSectorCompetition(raw: unknown): UpstreamSectorCompetition | null {
  return unwrap<UpstreamSectorCompetition>(raw);
}


function parseNumber(v: unknown): number | null {
  if (v == null || v === '') return null;
  const n = typeof v === 'number' ? v : Number(String(v).replace(/[, ]/g, ''));
  return Number.isFinite(n) ? n : null;
}


function peerMultipleFromBenchmark(
  sc: UpstreamSectorCompetition | null,
  metricCode: string,
): PeerMultipleStats | null {
  if (!sc?.benchmarks) return null;
  const b = sc.benchmarks.find(x => x.metric_code === metricCode);
  if (!b) return null;
  const median = b.median ?? null;
  const cv = b.company_value ?? null;
  const vsMedian = (cv != null && median != null && median !== 0)
    ? Math.round(((cv / median) - 1) * 10000) / 10000
    : null;
  return {
    metric: metricCode,
    count: b.company_value != null ? 1 : 0,
    median,
    q1: b.q1 ?? null,
    q3: b.q3 ?? null,
    company_value: cv,
    company_vs_median: vsMedian,
  };
}


export function adaptValuationForLegacy(
  fa: UpstreamFinancialAnalysis | null,
  sc: UpstreamSectorCompetition | null,
  ticker: string,
  outputId: string,
  opts: { llmMarkdownSource?: string | null } = {},
): LegacyValuationOutput {
  const warnings: string[] = [];
  if (!fa) {
    warnings.push('No financial_analysis output — valuation cannot pull engine DCF');
  }

  // Sector resolution: structured first, then LLM markdown cues, then
  // ticker heuristic. Mixed-flag mode (valuation Python + fa LLM)
  // leaves fa=null here; we still want banking/holding warnings to
  // fire so the LLM doesn't produce an FCF-DCF on a bank.
  const sectorResolution = resolveSector({
    structuredSector: fa?.sector ?? null,
    markdownSource: opts.llmMarkdownSource ?? null,
    ticker,
  });
  const sector = sectorResolution.sector;
  if (sectorResolution.source !== 'structured') {
    warnings.push(`Sector inferred via ${sectorResolution.source} fallback → ${sector}; LLM should verify.`);
  }
  const engine = fa?.engine_snapshot ?? fa?.engine ?? null;
  const dcf = engine?.dcf ?? null;

  const notes: string[] = [];
  let tryWacc = false;
  let holdingSotp = false;
  let bankingWarn = false;

  // TRY WACC trap — engine emits wacc_used as a decimal (0.32 = 32%);
  // anything above 25% looks like a TRY rate and the DCF will be
  // suppressed to near-zero. Warn and recommend USD WACC.
  if (dcf?.wacc_used != null) {
    const w = parseNumber(dcf.wacc_used);
    if (w != null && w > 0.25) {
      tryWacc = true;
      notes.push(`WACC used ${w.toFixed(3)} looks like a TRY rate. Prefer USD WACC for Turkish filers — TRY WACC trap collapses DCF.`);
    }
  }

  if (sector === 'holding') {
    holdingSotp = true;
    notes.push('Holding filer — consolidated DCF is an upper bound only. SOTP table required for real fair value.');
  }

  if (sector === 'banking') {
    bankingWarn = true;
    notes.push('Banking filer — FCF-based DCF not applicable. Prefer excess return / DDM. Engine DCF (if any) should be ignored.');
  }

  if (!dcf) {
    warnings.push('No engine DCF in upstream — LLM layer must run the DCF itself or supply valuation_inputs to the engine.');
  }

  const peerEvEbitda = peerMultipleFromBenchmark(sc, 'EBITDA_MARGIN');
  const peerPe       = peerMultipleFromBenchmark(sc, 'NET_MARGIN');

  return {
    agent_id: 'valuation_agent',
    output_id: outputId,
    ticker: (fa?.ticker ?? ticker).toUpperCase(),
    period_label: fa?.period_label ?? 'unknown',
    sector,
    dcf: bankingWarn ? null : dcf,  // refuse to surface an FCF-DCF on a bank
    peer_ev_ebitda: peerEvEbitda,
    peer_pe: peerPe,
    try_wacc_warning: tryWacc,
    holding_sotp_required: holdingSotp,
    banking_sector_warning: bankingWarn,
    sector_source: sectorResolution.source,
    notes,
    warnings,
    review_status: 'pending_ceo_review',
    source: 'python',
  };
}

File: python-services/src/financex/calculators/valuation.py
"""valuation — hybrid runner on top of financial_engine.

Python does the DCF arithmetic + peer-multiple aggregation. The LLM
layer picks WACC / terminal-growth policy, writes the Bull/Base/Bear
scenarios and composes the weighted fair value. TRY WACC is intentionally
flagged as unsafe on Turkish filers (see valuation_astor_learnings memory).
"""

from __future__ import annotations

import statistics
from dataclasses import dataclass
from datetime import UTC, datetime
from decimal import Decimal

from financex.calculators.financial_engine import (
    PiotroskiDeltas,
    ValuationInputs,
    compute_for_period,
)
from financex.schemas.analysis import FinancialAnalysisOutput
from financex.schemas.base import FinancexModel, Sector
from financex.schemas.engine import DcfResult
from financex.schemas.financials import PeriodFinancials


# ---------- Schemas ----------

class PeerMultipleStats(FinancexModel):
    metric: str
    count: int
    median: Decimal | None = None
    low: Decimal | None = None
    high: Decimal | None = None
    q1: Decimal | None = None
    q3: Decimal | None = None


class ValuationOutput(FinancexModel):
    ticker: str
    sector: Sector
    period_label: str

    dcf: DcfResult | None = None
    peer_ev_ebitda: PeerMultipleStats | None = None
    peer_pe: PeerMultipleStats | None = None

    # LLM hints
    try_wacc_warning: bool = False
    holding_sotp_required: bool = False
    banking_sector_warning: bool = False

    notes: list[str] = []
    computed_at: datetime
    schema_version: str = "1.0.0"


# ---------- Peer multiples ----------

@dataclass(frozen=True)
class _PeerMetric:
    code: str
    values: list[Decimal]


def _peer_stats(code: str, peers: list[FinancialAnalysisOutput]) -> PeerMultipleStats:
    values: list[Decimal] = []
    for p in peers:
        hv = p.canonical_numbers.get(code)
        if hv is not None:
            values.append(hv)
    if not values:
        return PeerMultipleStats(metric=code, count=0)
    sorted_vals = sorted(values)
    floats = [float(v) for v in values]
    q = statistics.quantiles(floats, n=4, method="inclusive") if len(floats) >= 2 else [floats[0]] * 3
    return PeerMultipleStats(
        metric=code,
        count=len(values),
        low=sorted_vals[0],
        high=sorted_vals[-1],
        q1=Decimal(str(round(q[0], 4))),
        median=Decimal(str(round(q[1], 4))),
        q3=Decimal(str(round(q[2], 4))),
    )


# ---------- Orchestrator ----------

def run_valuation(
    ticker: str,
    pf: PeriodFinancials,
    *,
    valuation_inputs: ValuationInputs | None = None,
    market_cap: Decimal | None = None,
    shares_outstanding: Decimal | None = None,
    peers: list[FinancialAnalysisOutput] | None = None,
) -> ValuationOutput:
    """Full valuation pass for one ticker.

    Runs the engine's DCF if valuation_inputs + shares supplied, then
    aggregates peer multiples where peers are given. Sector-aware: bank
    DCF would need FCFE + NIM assumptions — current engine does an
    industrial-style FCF DCF, so we flag the limitation instead of
    silently computing a misleading number.
    """
    notes: list[str] = []

    # Warn on sector mismatches
    try_wacc_warning = False
    holding_sotp = False
    banking_warning = False
    if valuation_inputs and valuation_inputs.wacc_override is not None:
        w = valuation_inputs.wacc_override
        # Turkish TRY WACC is typically 30-40%+; USD-equivalent 12-18%.
        if w > Decimal("0.25"):
            try_wacc_warning = True
            notes.append(
                f"WACC override {w} looks like a TRY rate. Memory warns this is the 'TRY WACC trap' — "
                "prefer USD WACC for Turkish filers unless FCFs are TRY-nominal and matched."
            )
    if pf.sector == Sector.HOLDING:
        holding_sotp = True
        notes.append("Holding — consolidated DCF is an upper bound only. SOTP required for real fair value.")
    if pf.sector == Sector.BANKING:
        banking_warning = True
        notes.append("Banking — FCF-based DCF not applicable; prefer excess return or DDM. Skipping DCF.")

    # DCF (industrial / holding only)
    dcf_result: DcfResult | None = None
    if pf.sector != Sector.BANKING and valuation_inputs:
        engine = compute_for_period(
            pf,
            market_cap=market_cap,
            shares_outstanding=shares_outstanding,
            valuation=valuation_inputs,
        )
        dcf_result = engine.dcf

    # Peer multiples
    peer_ev_ebitda = None
    peer_pe = None
    if peers:
        peer_ev_ebitda = _peer_stats("ebitda_margin", peers)  # proxy for EV/EBITDA relative pos
        peer_pe = _peer_stats("net_margin", peers)            # proxy for P/E relative pos

    return ValuationOutput(
        ticker=ticker.upper(),
        sector=pf.sector,
        period_label=f"{pf.period.value}-{pf.year}",
        dcf=dcf_result,
        peer_ev_ebitda=peer_ev_ebitda,
        peer_pe=peer_pe,
        try_wacc_warning=try_wacc_warning,
        holding_sotp_required=holding_sotp,
        banking_sector_warning=banking_warning,
        notes=notes,
        computed_at=datetime.now(UTC),
    )


# Convenience re-exports so callers don't need two imports.
__all__ = ["ValuationInputs", "PiotroskiDeltas", "ValuationOutput", "PeerMultipleStats", "run_valuation"]
